documentation/multstep-form.md

```diff
@@ -0,0 +1,243 @@
+form.first_name = 'Stan'
+form.company_id #=> 'Stan'
+```
+
+All the multistep form comes with a built-in attribute `completed_steps` which is relied upon by many of the internal methods.
+
+`#attributes` method is later on use to save the state of the form - see controller section of this doc.
+
+
+Each step is represented by a standalone form object and can be accessed via `steps` method:
+
+```ruby
+form.steps[:name].attributes #=> { first_name: 'Stan', last_name: nil }
+form.steps[:name].valid? #=> false
+```
+
+You can also access the step classes via `MyMultistepForm.steps`:
+
+```ruby
+MyMultistepForm.steps #=> { name: MyMultipleForm::Name, contact: MyMultipleForm::Contact, another_step: MyMultipleForm::AnotherStep }
+```
+
+##### Arrays
+In addition to standard `ActiveModel.attribute` usage, you can also denote attribute to be an array. This will:
+* sets `[]` as a default
+* automatically reject all blank values on assignment
+
+TODO: `attribute` method accepts typecasts argument as well, which is currently ignored for arrays. Would be nice to be able to declare attributes like:
+
+```ruby
+attribute :numbers, :integer, array: true # This currently ignored :integer
+```
+
+### Flow control
+
+Unlike most of the multistep forms libraries, we make very little assumptions about the order of the steps. The flow is defined by the `#next_step` method, which determines the next step of the form based on the current state of the form.
+The counterpoint to `next_step` method is `complete_step!` which updates the state of the form. 
+The default implementation of `next_step` methods simply assumes the linear order of the steps, in the order of their definitions:
+
+```ruby
+form = MyMultistepForm.new
+form.next_step #=> :name
+form.complete_step!(:name)
+
+form.next_step #=> :contact
+```
+
+However, it is important to note that steps can be completed in any order. Form will find its way around this:
+
+```ruby
+form = MyMultistepForm.new
+form.complete_step!(:contact)
+form.next_step #=> :name
+form.complete_step!(:name)
+form.next_step #=> :another_step
+```
+
+This means, we can override the `next_step` method to handle more interesting flows, like random order below:
+
+```ruby
+class MyMultistepForm
+  def next_step(current_step: nil, **)
+    super if current_step.present?
+    
+    @next_step ||= (self.class.steps.keys - completed_steps.keys).sample
+  end
+  
+  def complete_step!(*)
+    super
+    @next_step = nil
+  end
+end
+ 
+```
+NOTE: In above example we have fixed next_step so it returns the same, albait random, value of the next step until the step is completed.
+NOTE2: next_step takes an optional `current_step:` keyword. This one is used to return the step occurring after already completed step and its result depends on order of `complete_step!` calls. So the example above will perform all the steps in a random order, but will maintain a linear history once of the completed steps:
```

Comment:
From the example above and this note, I understand that what you've done ensures that the ordering of steps in `completed_steps` is determined by the order in which they were actually completed, is that correct?

If so, I'm not 100% clear on how this is based on the note above.

Reply:
That's correct :)

Reply:
I have re-worded that note.